src/main/java/com/aws/greengrass/dependency/Context.java

```diff
@@ -483,17 +494,21 @@ private T constructObjectWithInjection() {
                     int paramCount = pickedConstructor.getParameterCount();
                     if (paramCount == 0) {
                         // no arg constructor
-                        return putAndInjectFields(pickedConstructor.newInstance());
+                        return pickedConstructor.newInstance();
                     }
 
                     Object[] args = getOrCreateArgInstances(clazz, pickedConstructor, paramCount);
-                    return putAndInjectFields(pickedConstructor.newInstance(args));
+                    return pickedConstructor.newInstance(args);
                 } catch (Throwable ex) {
                     throw new IllegalArgumentException("Can't create instance of " + targetClass.getName(), ex);
                 }
             }
         }
 
+        private T constructObjectWithInjection() {
+            return putAndInjectFields(constructObject());
```

Comment:
This method should be guarded by the lock as well so that two threads calling `Value::get` can't both construct their own objects (one after the other) and then inject them both (one after the other).

Locking would work here because constructObject has the check and quick return on line 479, as well as putAndInjectFields on line 459

Reply:
Updated